src/System.Private.CoreLib/shared/System/Number.Grisu3.cs

```diff
@@ -0,0 +1,718 @@
+                    }
+
+                    case 16:
+                    case 15:
+                    case 14:
+                    {
+                        if (Ten4 <= number)
+                        {
+                            power = Ten4;
+                            exponent = 4;
+                            break;
+                        }
+
+                        goto case 13;
+                    }
+
+                    case 13:
+                    case 12:
+                    case 11:
+                    case 10:
+                    {
+                        if (1000 <= number)
+                        {
+                            power = 1000;
+                            exponent = 3;
+                            break;
+                        }
+
+                        goto case 9;
+                    }
+
+                    case 9:
+                    case 8:
+                    case 7:
+                    {
+                        if (100 <= number)
+                        {
+                            power = 100;
+                            exponent = 2;
+                            break;
+                        }
+
+                        goto case 6;
+                    }
+
+                    case 6:
+                    case 5:
+                    case 4:
+                    {
+                        if (10 <= number)
+                        {
+                            power = 10;
+                            exponent = 1;
+                            break;
+                        }
+
+                        goto case 3;
+                    }
+
+                    case 3:
+                    case 2:
+                    case 1:
+                    {
+                        if (1 <= number)
+                        {
+                            power = 1;
+                            exponent = 0;
+                            break;
+                        }
+
+                        goto case 0;
+                    }
+
+                    case 0:
+                    {
+                        power = 0;
+                        exponent = -1;
+                        break;
+                    }
+
+                    default:
+                    {
+                        power = 0;
+                        exponent = 0;
+
+                        Debug.Fail("unreachable");
+                        break;
+                    }
+                }
+            }
+
+            private static void CachedPower(int k, out DiyFp cmk, out int decimalExponent)
+            {
+                int index = ((PowerOffset + k - 1) / PowerDecimalExponentDistance) + 1;
+                cmk = new DiyFp(CachedPowerSignificands[index], CachedPowerBinaryExponents[index]);
+                decimalExponent = CachedPowerDecimalExponents[index];
+            }
+
+            private static bool DigitGen(ref DiyFp mp, int precision, char* digits, out int length, out int k)
```

Comment:
This is the function that is causing the most perf diff between the managed and native implementations.

We could likely close the gap by caching some field/properties we lookup multiple times (into locals), and by using `Math.DivRem` (or better, a proper intrinsic), rather than independent divide and remainder operations.